proposal: Shaper pipeline component

vblagoje opened the pull request:
### Proposed Changes:
Add a proposal for InputOutputShaper, aka Shaper pipeline component needed to make the PromptNode work more flexibly. 

### How did you test it?
N/A

### Notes for the reviewer
N/A

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I've used the [conventional commit convention](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

julian-risch commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+name of the variable in the invocation context that we want to expand. The `size`parameter is a result of the `len`
+function invocation on the variable `documents`.
+
+For the documents variable, we want to invoke the function `contract_docs` and store the result in the same variable.
```
`contract_docs` might be confusing here. What about calling this function `concat` and adding an `in_place` parameter? That parameter could make all the functions more flexible. No need to introduce `contract_docs` and `contract_...` if we need the concatenation in future not only for documents. 
The `expand` function (and any other future functions) could also get an `in_place` parameter.

vblagoje replied:
`concat` sounds great but what is the objective of `in_place`?

julian-risch replied:
The `in_place` or `inplace` parameter would make setting `output` parameter obsolete if there is no renaming of the input. We could reuse the concat function (and other functions) and have `in_place` set to `True` by default. 

Right now, `contract_docs` can't be easily reused let's say for answers because the default output is docs. My suggestion is to rename the function to just `concat` and have a `in_place` or `inplace` set to `True` by default, which should have the effect that the output is the same as the input. `documents` if it's used on `documents` and `answers` if it's used on `answers`. 

I got this thought because I am working with pandas right now. Something like `drop_duplicates(inplace=True) `https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html
Whether that's a good idea for Haystack depends on whether many functions will be reused for different inputs (documents, answers, etc.) and whether we always rename the output or sometimes have the same output name as the input.

vblagoje replied:
Aha gotcha, this was already in the implementation, but I forgot to mention it here. If there is an output parameter specified - then it is used, otherwise, the value of the invocation is assigned to the input variable. I'll add a section on "Ommiting output parameter"

julian-risch commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+
+For example, in the YAML snippet definition below, we have a function `expand` that takes two keyword parameters:
+`expand_target` and `size`. However, we haven't specified either of these parameters in the YAML config. This is
+possible because we assume that the first parameter is always the variable we want to invoke the function on. In this
```
instead of this assumption, the `expand` function and all other shaper functions could have an `in_place` parameter that is set to True by default.

julian-risch commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+# Unresolved questions
+
+Optional, but suggested for first drafts. What parts of the design are still
+TBD?
```
Let's not forget about the `run_batch` methods of the new node. I don't foresee any complications with that but so far it's not mentioned at all.

julian-risch commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+
+# Basic example
+
+In the following example, we use InputOutputShaper to add a new variable `questions` to the invocation context.
```
What about using the shorter name: `IOShaper` ?

vblagoje replied:
Excellent!

mathislucka replied:
why use IO at all? I think this will cause confusions because people usually connect IO with disk or network read/write operations.

vblagoje replied:
Ok interesting point. So simply Shaper then?

mathislucka replied:
Yes I'd go for Shaper then.

dmigo commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+
+# Alternatives
+
+A better solution would likely be a more general run method for components. This would allow us to arbitrarily
+define the pipeline invocation context. However even in those cases we'll need to use InputOutputShaper to modify
+existing variables in invocation context as needed.
```
It seems like we are about to create a data transformation language of our own here. This is a big effort. Have we considered using an existing language, eg https://getdtl.org/ ?

vblagoje replied:
Cool, is there a DTL reference impl for Python?

vblagoje replied:
But to expand - we need to sometimes rename/duplicate vars, and invoke some arbitrary function on the input data. There is no goal to create DTL or enable some fancy pre/post-processing capabilities for the pipeline component.

mkkuemmel commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+- Title: Input/Output Shaper
+- Decision driver: Vladimir
+- Start Date: 2023-12-29
```
2022-12-29

vblagoje replied:
:-) good catch

mkkuemmel commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+
+# Detailed design
+
+The InputOutputShaper component is primarily used in pipelines and configured via a YAML file. The YAML component
```
Why "primarily"? Where else would it be used?

vblagoje replied:
I meant to say in pipeline definitions; perhaps the wording is off.  But you are right, noted!

mkkuemmel commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+name of the variable in the invocation context that we want to expand. The `size`parameter is a result of the `len`
+function invocation on the variable `documents`.
+
+For the documents variable, we want to invoke the function `contract_docs` and store the result in the same variable.
```
I find the naming `contract_docs` slightly confusing, too (thinking of contracts).

vblagoje replied:
Yeah, Julian was spot on as [well](https://github.com/deepset-ai/haystack/pull/3784#discussion_r1059899887)

mkkuemmel commented on proposals/text/3784-input-output-shaper.md:
```diff
@@ -0,0 +1,247 @@
+- InputOutputShaper could turn out to be useful in other use cases as well - i.e. declarative pre/post processing.
+- InputOutputShaper integrates well with PromptNodes and other components.
+- No braking changes to existing components.
+
```
Most of these points are not drawbacks :)

vblagoje replied:
Thank @mkkuemmel - I answered questions in the proposal template mindlessly. Noted.

vblagoje commented:
Thanks @dmigo @mkkuemmel @mathislucka @julian-risch, for the first round of comments. I renamed the component to simply `Shaper`. I also added a section regarding the option to omit the `output` function parameter. We can now focus on details if there are no other significant omissions.

mathislucka commented:
One question @vblagoje what actually happens to document meta and document ids when we concat multiple documents?

vblagoje commented:
> One question @vblagoje what actually happens to document meta and document ids when we concat multiple documents?

Whatever we need to do will happen. I left this out of proposal as we can iterate the registered functions through releases as we see fit.

julian-risch reviewed (approved):
We can talk about how to implement `run_batch` later. For the first version, simply calling `run` iteratively would be fine but I am sure we can do better than that. 🙂

vblagoje commented:
@dmigo @mkkuemmel would you please review the changes I made; if you don't see any more issues - please approve the proposal.

mayankjobanputra commented:
I think this is much needed component in Haystack to support PromptNode through YAML pipelines. I have a few suggestions as below:

### Suggestions:

- I don’t fully support the node “Shaper” as we use it for non-shaping tasks such as “rename”. Maybe we should call it “Transducer”? and make the design even more flexible (won’t implement everything from the beginning though). Reasons below.
- I am also not sure if we should assume any default parameters as Shaper is for mostly for the YAML users. I suggest we provide templates but make user define the whole shaper in the YAML.

#### Reasons:
1. Today PromptNode only support "flan"-like models. I am expecting a little redesign of the PromptNode soon to support [instructor-embedding](https://github.com/HKUNLP/instructor-embedding)
2. Although today we only support a handful of Shaping functions (i.e., rename, expand, concat) and this design may suffice for current Templates and use cases, if prompting picks up traction, YAML users might demand more transformations.


### Question:
1. How can user expand Shaper to add new features? Is there a solution that doesn't require adding a custom node?

vblagoje commented:
> 

@mayankjobanputra I am indifferent to what the component's name ends up being. I don't think we should spend much time debating the name of a component which is not the core of the project offering. This is a helper component that most people will not use. For the default parameters, let's ask @mathislucka because his reasoning is opposite to yours. And I agree with him tbh. dC doesn't want to have verbose config files if the same use cases repeatedly appear. The registered functions of Shaper are not set in stone, and they can evolve. We'll surely add more functions in the future without changing the inner workings of the Shaper. And that was the point. Keep the component logic and evolve functions.
